pkg/agent/memberlist/cluster.go

```diff
@@ -0,0 +1,628 @@
+	conHashMapRWLock sync.RWMutex
+	// nodeEventsCh, the Node join/leave events will be notified via it.
+	nodeEventsCh chan memberlist.NodeEvent
+
+	// ClusterNodeEventHandlers hold the fun that should implement when a node join or leave the cluster
+	// for example, when a nwe node join the cluster,
+	// each node should recalculate that if it should still hold the existing egresses owned by itself
+	// in the same way, when a node leave cluster, each node may get some new egresses from the left node
+	ClusterNodeEventHandlers []clusterNodeEventHandler
+
+	// existingMembers contains all node(ip:clusterPort) that should join memberlist cluster
+	// the items list from nodelist and filter with externalIPPool nodeSelectors
+	// if the local node is not joined Cluster, it should hold the nodes,
+	// once local node should join cluster, it will join cluster with those nodes
+	memberRWLock    sync.RWMutex
+	existingMembers sets.String
+
+	localNodeEventCh chan bool
+	conHashEventCh   chan clusterHashEvent
+
+	nodeInformer     coreinformers.NodeInformer
+	nodeLister       corelisters.NodeLister
+	nodeListerSynced cache.InformerSynced
+
+	ipPoolInformer          crdinformers.ExternalIPPoolInformer
+	ipPoolLister            crdlister.ExternalIPPoolLister
+	ipPoolInformerHasSynced cache.InformerSynced
+}
+
+// NewCluster return a new *Cluster
+func NewCluster(
+	clusterBindPort int,
+	nodeInformer coreinformers.NodeInformer,
+	nodeConfig *config.NodeConfig,
+	ipPoolInformer crdinformers.ExternalIPPoolInformer,
+) (*Cluster, error) {
+	// The Node join/leave events will be notified via it.
+	nodeEventCh := make(chan memberlist.NodeEvent, 1024)
+	c := &Cluster{
+		bindPort:                clusterBindPort,
+		NodeConf:                nodeConfig,
+		nodeName:                nodeConfig.Name,
+		nodeEventsCh:            nodeEventCh,
+		nodeInformer:            nodeInformer,
+		nodeLister:              nodeInformer.Lister(),
+		nodeListerSynced:        nodeInformer.Informer().HasSynced,
+		ipPoolInformer:          ipPoolInformer,
+		ipPoolLister:            ipPoolInformer.Lister(),
+		ipPoolInformerHasSynced: ipPoolInformer.Informer().HasSynced,
+		localNodeEventCh:        make(chan bool, 1),
+		conHashEventCh:          make(chan clusterHashEvent, 1),
+	}
+	c.conHashMap = make(map[string]*consistenthash.Map)
+
+	bindPort := c.bindPort
+	hostIP := c.NodeConf.NodeIPAddr.IP
+	nodeMember := fmt.Sprintf("%s:%d", hostIP.String(), bindPort)
+	klog.V(2).InfoS("Add new node", "local node", nodeMember)
+
+	conf := memberlist.DefaultLocalConfig()
+	conf.Name = c.nodeName
+	conf.BindPort = bindPort
+	conf.AdvertisePort = bindPort
+	conf.Events = &memberlist.ChannelEventDelegate{Ch: nodeEventCh}
+	klog.V(1).Infof("Memberlist cluster configs: %+v", conf)
+
+	mList, err := memberlist.Create(conf)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create memberlist: %v", err)
+	}
+
+	c.mlConfig = conf
+	c.mList = mList
+	c.existingMembers = sets.NewString(nodeMember)
```

Comment:
From the comment of memberlist's `Join` method, it will try to contact all the given hosts, so I wonder whether it's useful to add the node itself. Have you tried it?

Reply:
Yes, you are right